src/main/java/ch/njol/skript/effects/EffChange.java

```diff
@@ -280,38 +281,40 @@ else if (mode == ChangeMode.SET)
 	}
 	
 	@Override
-	protected void execute(Event e) {
-		Object[] delta = changer == null ? null : changer.getArray(e);
-		delta = changer == null ? delta : changer.beforeChange(changed, delta);
+	protected void execute(Event event) {
+		Object[] deltaValues = null;
+		if (deltaValuesExpression != null) {
+			deltaValues = deltaValuesExpression.getArray(event);
+			deltaValues = deltaValuesExpression.beforeChange(expressionToChange, deltaValues);
+		}
 
-		if ((delta == null || delta.length == 0) && (mode != ChangeMode.DELETE && mode != ChangeMode.RESET)) {
-			if (mode == ChangeMode.SET && changed.acceptChange(ChangeMode.DELETE) != null)
-				changed.change(e, null, ChangeMode.DELETE);
+		boolean deltaValuesPresent = deltaValues != null && deltaValues.length != 0;
+		boolean deltaValuesRequired = changeMode != ChangeMode.DELETE && changeMode != ChangeMode.RESET;
+		if (deltaValuesRequired && !deltaValuesPresent) {
+			boolean expressionToChangeAllowsDeletion = expressionToChange.acceptChange(ChangeMode.DELETE) != null;
+			if (changeMode == ChangeMode.SET && expressionToChangeAllowsDeletion)
+				expressionToChange.change(event, null, ChangeMode.DELETE);
 			return;
 		}
-		changed.change(e, delta, mode);
+		expressionToChange.change(event, deltaValues, changeMode);
 	}
 	
 	@Override
-	public String toString(final @Nullable Event e, final boolean debug) {
-		final Expression<?> changer = this.changer;
-		switch (mode) {
+	public String toString(final @Nullable Event event, final boolean debug) {
+		assert deltaValuesExpression != null;
```

Comment:
Won't this cause problems in test mode if it ever gets stringified? For `DELETE` and `RESET` it is null, right? (If not the field probably shouldn't be `@Nullable` so we won't need this assertion)